R/util_dfc.R

```diff
@@ -0,0 +1,238 @@
+    util.print_pre(length(unique_id_compare_list$shared_elements) %+% " shared unique identifiers, such as: " %+%
+                     paste(head(unique_id_compare_list$shared_elements), collapse = ", "))
+    util.print_pre(length(unique_id_compare_list$only_in_first) %+% " unique identifiers only in first data frame.")
+    if(length(unique_id_compare_list$only_in_first) > 0) {
+      util.print_pre("...such as: " %+% paste(head(unique_id_compare_list$only_in_first), collapse = ", "))}
+    util.print_pre(length(unique_id_compare_list$only_in_second) %+% " unique identifiers only in second data frame.")
+    if(length(unique_id_compare_list$only_in_second) > 0) {
+      util.print_pre("...such as: " %+% paste(head(unique_id_compare_list$only_in_second), collapse = ", "))}
+
+    # Use the overlap to filter both data frames to rows with shared IDs, if possible
+    if(length(unique_id_compare_list$shared_elements) == 0) {
+      util.print_pre("No shared IDs across data frames, so no way to compare data frame values. Ending function.")
+      return()
+    }
+    df1 <- df1[df1$temp_util_id %in% unique_id_compare_list$shared_elements, ]
+    df2 <- df2[df2$temp_util_id %in% unique_id_compare_list$shared_elements, ]
+    util.print_pre("Both data frames filtered to remove any rows with non-shared IDs. Any duplicate IDs within a data frame are preserved.")
+
+    # Sort both DFs by ID, and warn user if duplicates make perfect sorting impossible
+    df1 <- arrange(df1, temp_util_id)
+    df2 <- arrange(df2, temp_util_id)
+    util.print_pre("Rows of both data frames sorted by ID column(s).")
+    if(dup_IDs) {
+      util.print_pre("WARNING - ID columns do not uniquely identify rows in at least one data frame, so the rows of the two data frames cannot be guaranteed to match up.")
+    }
+  } else {
+    util.print_pre("WARNING - no ID columns were defined in the function call, so no way to filter or sort rows for comparison.")
+  }
+
+
+  ##### INDIVIDUAL-VALUE COMPARISON
+
+  # Sanity-check that the dimensions of the two DFs are the same
+  if(!all(dim(df1) == dim(df2))){
+    util.print_pre("WARNING - data frames do not have same dimensions after attempting to filter to shared rows and columns. No way to compare values. Stopping function.")
+    return()
+  }
+
+  # Actually compare the two data frames and tabulate the number of matches for each column.
+  # First, cast NAs to the string "NA" so that "NA" == "NA" returns TRUE.
+  df1[is.na(df1)] <- "NA"
+  df2[is.na(df2)] <- "NA"
+  # Also cut the temp_until_id column as you do the comparison.
+  dfdiff <- as.data.frame(df1[, !names(df1) %in% "temp_util_id"] == df2[, !names(df2) %in% "temp_util_id"])
+  ds.summarize_by_column(dfdiff, func_list = list("pct_matched" = mean)) %>%
+                           util.html_table()
+  if(all(dfdiff == TRUE)) {
+    util.print_pre("After subsetting and sorting on shared columns (by name) and rows (by ID, if possible), the data frames match.")
+  }
+
+}
+
```

Comment:
I expected two additional pieces of information:
1. List (and count) of column names for which the values are not identical, and how many non-matching values. For example, the output can be something like column A (12 non-matches), column C (2 non-matches), etc.... If more than 5 columns differ, you can truncate them.
2. List (and count) of ids which do not match. For example studen_a_week_3 (4 non-matches), student_b_week_1 (2 non_matches). Again, truncate if more than 5.

Reply:
#1 makes sense, thanks! #2 I think I do in the earlier section that starts with "COMPARING IDENTIFIERS: "

Reply:
Actually I think I do something close to Point 1 already when I apply ds.summarize_by_column() to the data frame dfdiff. I'll add a function to count non-matches.